lib/axon/losses.ex

```diff
@@ -958,6 +958,49 @@ defmodule Axon.Losses do
     )
   end
 
+  @doc ~S"""
+  Cosine Similarity error loss function.
+
+  $$l_i = \sum_i (\hat{y_i} - y_i)^2$$
+
+  ## Argument Shapes
+
+    * `y_true` - $(d_0, d_1, ..., d_n)$
+    * `y_pred` - $(d_0, d_1, ..., d_n)$
+
+  ## Options
+
+    * `:axes` - Defaults to `[1]`.
+    * `:eps` - Defaults to `0.000001`.
+
+  ## Examples
+      iex> y_pred = Nx.tensor([[1.0, 0.0], [1.0, 1.0]])
+      iex> y_true = Nx.tensor([[0.0, 1.0], [1.0, 1.0]])
+      #Nx.Tensor<
+        f32[2]
+        [0.0, 1.0]
+      >
+
+  ## Implemented in pytorch
+  https://github.com/pytorch/pytorch/blob/23902fb8955d48b8d1dfdd54c40e72c2744cac8e/aten/src/ATen/native/Distance.cpp#L242
+  """
+
+  defn cosine_similarity(y_true, y_pred, opts \\ []) do
+    opts = keyword!(opts, axes: [1], eps: 0.000001)
+
+    transform(
+      {opts[:axes], opts[:eps]},
+      fn
+        {axes, eps} ->
+          w12 = Nx.sum(y_true * y_pred, axes: axes)
+          w1 = Nx.sum(y_true * y_true, axes: axes)
+          w2 = Nx.sum(y_pred * y_pred, axes: axes)
+          n12 = Nx.sqrt(Nx.max(w1 * w2, eps * eps))
```

Comment:
```suggestion
          n12 = Nx.max(w1 * w2, eps)
```

Reply:
Perhaps its closer to the KERAS implementation then?
```elixir
  defn loss_cosine_similarity_keras(y_true, y_pred) do
    #https://github.com/keras-team/keras/blob/89879e2c76e86c685e44c47a6cdb82f7e645c142/keras/losses.py#L2167
    #y_true = tf.linalg.l2_normalize(y_true, axis=axis)
    #y_pred = tf.linalg.l2_normalize(y_pred, axis=axis)
    #return -tf.reduce_sum(y_true * y_pred, axis=axis)

    y_pred_l2 = Nx.divide(y_pred, Nx.LinAlg.norm(y_pred, axes: [1]) |> Nx.new_axis(1))
    y_true_l2 = Nx.divide(y_true, Nx.LinAlg.norm(y_true, axes: [1]) |> Nx.new_axis(1))
    -Nx.sum(y_true_l2 * y_pred_l2) / 2
  end
```
But im not sure to what degree these are equivalent (the PR returns a tensor, this one returns a number)